File: scripts/macos/ocr.m
#import <Foundation/Foundation.h>
#import <CoreImage/CoreImage.h>
#import <Vision/Vision.h>

// Function declarations
CGImageRef convertCIImageToCGImage(CIImage *inputImage);
void recognizeText(NSURL *fileName, NSArray<NSString *> *recognitionLanguages);
BOOL detectQRCode(NSURL *fileName);

int main(int argc, const char * argv[]) {
    @autoreleasepool {
        if (@available(macOS 10.15, *)) {
            if (argc < 2) {
                NSLog(@"Usage: %s <imagePath> [languages]", argv[0]);
                return 1;
            }
            
            NSString *imagePath = [NSString stringWithUTF8String:argv[1]];
            NSMutableArray<NSString *> *languages = [NSMutableArray arrayWithArray:@[@"zh-Hans", @"en-US"]];
            
            if (argc > 2) {
                [languages removeAllObjects];
                for (int i = 2; i < argc; i++) {
                    NSString *language = [NSString stringWithUTF8String:argv[i]];
                    [languages addObject:language];
                }
            }

            NSURL *imageURL = [NSURL fileURLWithPath:imagePath];
            if (![[NSFileManager defaultManager] fileExistsAtPath:imagePath]) {
                NSLog(@"Invalid image path");
                return 1;
            }

            if (!detectQRCode(imageURL)) {
                recognizeText(imageURL, languages);
            }
        } else {
            NSLog(@"Only support macOS 10.15 and above");
        }
    }
    return 0;
}

CGImageRef convertCIImageToCGImage(CIImage *inputImage) {
    CIContext *context = [CIContext contextWithOptions:nil];
    return [context createCGImage:inputImage fromRect:inputImage.extent];
}

void recognizeText(NSURL *fileName, NSArray<NSString *> *recognitionLanguages) {
    if (@available(macOS 10.15, *)) {
        CIImage *ciImage = [CIImage imageWithContentsOfURL:fileName];
        CGImageRef img = convertCIImageToCGImage(ciImage);
        if (!ciImage || !img) {
            return;
        }
        
        VNImageRequestHandler *requestHandler = [[VNImageRequestHandler alloc] initWithCGImage:img options:@{}];
        
        VNRecognizeTextRequest *request = [[VNRecognizeTextRequest alloc] initWithCompletionHandler:^(VNRequest *request, NSError *error) {
            if (error) {
                fprintf(stderr, "Text recognition error: %s\n", error.localizedDescription.UTF8String);
                return;
            }

            NSArray<VNRecognizedTextObservation *> *observations = (NSArray<VNRecognizedTextObservation *> *)request.results;
            NSMutableArray<NSString *> *recognizedStrings = [NSMutableArray array];
            
            for (VNRecognizedTextObservation *observation in observations) {
                VNRecognizedText *topCandidate = [[observation topCandidates:1] firstObject];
                if (topCandidate) {
                    [recognizedStrings addObject:topCandidate.string];
                }
            }
            
            NSString *result = [recognizedStrings componentsJoinedByString:@" "];
            printf("%s\n", result.UTF8String);
        }];
        
        request.recognitionLanguages = recognitionLanguages;
        
        NSError *error = nil;
        if (![requestHandler performRequests:@[request] error:&error]) {
            fprintf(stderr, "Unable to perform the requests: %s\n", error.localizedDescription.UTF8String);
        }
    }
}

BOOL detectQRCode(NSURL *fileName) {
    CIImage *ciImage = [CIImage imageWithContentsOfURL:fileName];
    if (!ciImage) {
        return NO;
    }

    CIDetector *detector = [CIDetector detectorOfType:CIDetectorTypeQRCode context:nil options:@{CIDetectorAccuracy: CIDetectorAccuracyHigh}];
    NSArray *features = [detector featuresInImage:ciImage];

    BOOL isQRCode = NO;
    NSMutableString *result = [NSMutableString string];
    for (CIFeature *feature in features) {
        if ([feature.type isEqualToString:CIFeatureTypeQRCode]) {
            CIQRCodeFeature *qrFeature = (CIQRCodeFeature *)feature;
            isQRCode = YES;
            [result appendString:qrFeature.messageString ?: @""];
            [result appendString:@"\n"];
        }
    }

    if (isQRCode) {
        printf("%s\n", result.UTF8String);
    }

    return isQRCode;
}

File: scripts/macos/ocr.swift
import Foundation
import CoreImage
import Vision
import ArgumentParser

let joiner = " "

// Define the arguments accepted by the program
struct OCRArguments: ParsableCommand {
  @Argument(help: "The path to the image file.")
  var imagePath: String

  @Option(help: "Optional language codes for text recognition. Separate by spaces (e.g., 'zh-Hans en-US').")
  var languages: [String] = ["zh-Hans", "en-US"]
}

func convertCIImageToCGImage(inputImage: CIImage) -> CGImage? {
  let context = CIContext(options: nil)
  return context.createCGImage(inputImage, from: inputImage.extent)
}

@available(OSX 10.15, *)
func recognizeTextHandler(request: VNRequest, error: Error?) {
  guard let observations = request.results as? [VNRecognizedTextObservation] else {
      return
  }
  let recognizedStrings = observations.compactMap { observation in
      // Return the string of the top VNRecognizedText instance.
      observation.topCandidates(1).first?.string
  }

  // Process the recognized strings.
  let result = recognizedStrings.joined(separator: joiner)

  print(result)
}

@available(OSX 10.15, *)
func detectText(fileName: URL, recognitionLanguages: [String]) {
  guard
          let ciImage = CIImage(contentsOf: fileName),
          let img = convertCIImageToCGImage(inputImage: ciImage)
  else {
      return
  }

  let requestHandler = VNImageRequestHandler(cgImage: img)

  // Create a new request to recognize text.
  let request = VNRecognizeTextRequest(completionHandler: recognizeTextHandler)
  request.recognitionLanguages = recognitionLanguages

  do {
      // Perform the text-recognition request.
      try requestHandler.perform([request])
  } catch {
      print("Unable to perform the requests: \(error).")
  }
}

func detectImage(fileName: URL) -> Bool {
  let detector = CIDetector(ofType: CIDetectorTypeQRCode, context: nil, options: nil)

  guard let ciImage = CIImage(contentsOf: fileName), let features = detector?.features(in: ciImage) else {
      return false
  }

  var isQRCode = false
  var result = ""
  for feature in features as! [CIQRCodeFeature] {
      if feature.type == "QRCode" {
          isQRCode = true
      }
      result += feature.messageString ?? ""
      result += "\n"
  }

  if isQRCode {
      print(result)
  }

  return isQRCode
}

func main() {
  // 只支持 macOS 10.15 以上
  guard #available(OSX 10.15, *) else {
    return print("Only support the system above macOS 10.15")
  }

    do {
        let arguments = try OCRArguments.parse()
        
        let imagePath = URL(fileURLWithPath: arguments.imagePath)
        
        if imagePath.path.isEmpty {
            print("读取图片路径出错")
            return
        }
        
        // 检测是否有二维码，直接将二维码转换成解码后的内容
        guard detectImage(fileName: imagePath) else {
            return detectText(fileName: imagePath, recognitionLanguages: arguments.languages)
        }
    } catch {
        print("解析命令行错误")
    }
}
